### PR TITLE
Support AWS-LC in FIPS mode

### PR DESCRIPTION
**Pull request by mulmarta**
Changes:
* FIPS does not have `X509_get0_pubkey` so it's replaced by `X509_get_pubkey` which differs in requiring `free` of the key
* It seems that aws-lc-rs updated `BIO_number_written` and now it returns `u64` instead of `usize`. This is not the case for FIPS.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT license.

### CODE REVIEW

**Review by tomleavy — DISMISSED** [review]

File: mls-rs-crypto-awslc/src/x509/certificate.rs [hunk 1]
```diff
@@ -197,16 +197,17 @@ impl Certificate {
 
     pub fn public_key(&self) -> Result<SignaturePublicKey, AwsLcCryptoError> {
         unsafe {
-            let pub_key = X509_get0_pubkey(self.0);
+            let pub_key = X509_get_pubkey(self.0);
             let ec_key = EVP_PKEY_get0_EC_KEY(pub_key);
+            EVP_PKEY_free(pub_key);
```
Comment (stefunctional) [inline mls-rs-crypto-awslc/src/x509/certificate.rs:202]: Is that done too early given that `ec_key` is used after? According to [this](https://docs.openssl.org/3.0/man3/EVP_PKEY_set1_RSA/):

> EVP_PKEY_get0_EC_KEY() return the referenced key in pkey or NULL if the key is not of the correct type. The reference count of the returned key is not incremented and so the key must not be freed after use.

Comment (mulmarta) [reply mls-rs-crypto-awslc/src/x509/certificate.rs:202]: Hm, I'm not sure what this does internally but wouldn't it be crashing if we freed before use? And it's not crashing.

Comment (stefunctional) [reply mls-rs-crypto-awslc/src/x509/certificate.rs:202]: A use-after-free does not necessarily cause a crash ([example](https://godbolt.org/z/4vE4943jW)). They can lead to vulnerabilities like remote code execution.

The documentation is not very clear (there are multiple pointers involved) and I have only glanced over the C code, so I am not sure yet if there really is an issue here. Also not sure if running a tool like valgrind or some memory sanitizer while exercising this code could help. I'll try to look into it more.

Comment (mulmarta) [reply mls-rs-crypto-awslc/src/x509/certificate.rs:202]: I can also free at the end to be on the safe side. I guess it's only slightly more annoying.

Comment (mulmarta) [reply mls-rs-crypto-awslc/src/x509/certificate.rs:202]: Also there's FIPS code and non-FIPS C code which may be different, although it's unlikely.

Comment (tomleavy) [reply mls-rs-crypto-awslc/src/x509/certificate.rs:202]: get0 means no reference count increase (return an unowned reference to the internal data). Freeing would be bad here as @stefunctional points out.

**Review by stefunctional — APPROVED** [review]

**Review by tomleavy — APPROVED** [review]

**Pull request merged**